util/chunk/codec.go

```diff
@@ -189,3 +189,139 @@ func init() {
 		allNotNullBitmap[i] = 0xFF
 	}
 }
+
+// Decoder decodes the data returned from the coprocessor and stores the result in Chunk.
+// How Decoder works:
+// 1. Initialization phase: Decode a whole input byte slice to Decoder.intermChk using Codec.Decode. intermChk is
+//    introduced to simplify the implementation of decode phase. This phase uses pointer operations with less CPU andF
+//    memory cost.
+// 2. Decode phase:
+//    2.1 Set the number of rows that should be decoded to a multiple of 8 greater than
+//        chk.RequiredRows() - chk.NumRows(), this can reduce the cost when copying the srcCol.nullBitMap into destCol.nullBitMap.
+//    2.2 Append srcCol.offsets to destCol.offsets when the elements is of var-length type. And further adjust the
+//        offsets according to descCol.offsets[destCol.length]-srcCol.offsets[0].
+//    2.3 Append srcCol.nullBitMap to destCol.nullBitMap.
+// 3. Go to step 1 when the input byte slice is consumed.
+type Decoder struct {
+	intermChk    *Chunk
+	colTypes     []*types.FieldType
+	remainedRows int
+}
+
+// NewDecoder creates a new Decoder object for decode a Chunk.
+func NewDecoder(chk *Chunk, colTypes []*types.FieldType) *Decoder {
+	return &Decoder{intermChk: chk, colTypes: colTypes, remainedRows: 0}
+}
+
+// Decode decodes multiple rows of Decoder.intermChk and stores the result in chk.
+func (c *Decoder) Decode(chk *Chunk) {
+	requiredRows := chk.RequiredRows() - chk.NumRows()
+	// Set the requiredRows to a multiple of 8.
+	requiredRows = (requiredRows + 7) >> 3 << 3
+	if requiredRows > c.remainedRows {
+		requiredRows = c.remainedRows
+	}
+	for i := 0; i < len(c.colTypes); i++ {
+		c.decodeColumn(chk, i, requiredRows)
+	}
+	c.remainedRows -= requiredRows
+}
+
+// Reset decodes data and store the result in Decoder.intermChk. This decode phase uses pointer operations with less
+// CPU and memory costs.
+func (c *Decoder) Reset(data []byte) {
+	codec := NewCodec(c.colTypes)
```

Comment:
Can `Codec.colTypes` be removed?
`colTypes` is only used when `getFixedLen` in decodeColumn, but we can get the `elemLen` from `cap(chunk.Column[ordinal].elemLen)`. `len(column.offsets) != 0` means `isVarType`?

Reply:
No. When we need adjust the Endian, the colTypes is necessary. 
But one of `Codec` and `Decoder` have `colTypes` is enough.  
I think  keep it in `Codec` is better.